plugins/genesys/fu-genesys-usbhub-firmware.c

```diff
@@ -223,75 +323,90 @@ fu_genesys_usbhub_firmware_parse(FuFirmware *firmware,
-						   &kbs,
-						   error))
+			if (!fu_genesys_usbhub_firmware_query_codesize(firmware, fw, offset, error))
 				return FALSE;
-			code_size = 1024 * kbs;
 		} else {
-			code_size = 0x6000;
+			fu_firmware_set_size(firmware, 0x6000);
 		}
 		break;
 	}
-	case ISP_MODEL_HUB_GL3590: {
-		guint8 kbs = 0;
-		if (!fu_memread_uint8_safe(buf,
-					   bufsz,
-					   GENESYS_USBHUB_CODE_SIZE_OFFSET,
-					   &kbs,
-					   error))
-			return FALSE;
-		code_size = 1024 * kbs;
-		break;
-	}
+	case ISP_MODEL_HUB_GL3590:
 	case ISP_MODEL_HUB_GL3525: {
-		guint8 kbs = 0;
-		if (!fu_memread_uint8_safe(buf,
-					   bufsz,
-					   GENESYS_USBHUB_CODE_SIZE_OFFSET,
-					   &kbs,
-					   error))
+		if (!fu_genesys_usbhub_firmware_query_codesize(firmware, fw, offset, error))
 			return FALSE;
-		code_size = 1024 * kbs;
 		break;
 	}
 	default:
 		break;
 	}
-	fu_firmware_set_size(firmware, code_size);
+	code_size = fu_firmware_get_size(firmware);
 
 	/* calculate checksum */
 	if ((flags & FWUPD_INSTALL_FLAG_IGNORE_CHECKSUM) == 0)
-		if (!fu_genesys_usbhub_firmware_verify(buf, bufsz, code_size, error))
+		if (!fu_genesys_usbhub_firmware_verify(fw, offset, code_size, error))
 			return FALSE;
 
 	/* get firmware version */
-	if (!fu_memread_uint16_safe(buf,
-				    bufsz,
-				    GENESYS_USBHUB_VERSION_OFFSET,
-				    &version_raw,
-				    G_LITTLE_ENDIAN,
-				    error))
+	if (!fu_genesys_usbhub_firmware_query_version(firmware, fw, offset, error))
 		return FALSE;
-	fu_firmware_set_version_raw(firmware, version_raw);
-	version =
-	    g_strdup_printf("%02x.%02x", (version_raw & 0xFF00U) >> 8, (version_raw & 0x00FFU));
-	fu_firmware_set_version(firmware, version);
+
+	/* parse remaining firmware bytes */
+	if (bufsz - code_size > 0) {
+		gsize fw_offset = offset + code_size;
+		gsize remaining = bufsz - code_size;
+
+		while (remaining) {
+			FuGenesysFwType fw_type =
+			    fu_genesys_usbhub_firmware_query_magic(fw, fw_offset, error);
+			g_autoptr(FuFirmware) firmware_sub = NULL;
+			g_autoptr(GBytes) blob = NULL;
+
+			switch (fw_type) {
+			case FU_GENESYS_FW_TYPE_HUB:
+				g_set_error_literal(error,
+						    FWUPD_ERROR,
+						    FWUPD_ERROR_NOT_SUPPORTED,
+						    "fw bytes have dual hub firmware");
+				return FALSE;
+			case FU_GENESYS_FW_TYPE_DEV_BRIDGE:
+				firmware_sub = fu_genesys_usbhub_dev_firmware_new();
+				break;
+			case FU_GENESYS_FW_TYPE_PD:
+				firmware_sub = fu_genesys_usbhub_pd_firmware_new();
+				break;
+			case FU_GENESYS_FW_TYPE_CODESIGN:
+				firmware_sub = fu_genesys_usbhub_codesign_firmware_new();
+				break;
+			default:
+				return FALSE;
+			}
+			/* parse firmware */
+			blob = fu_bytes_new_offset(fw, fw_offset, remaining, error);
+			if (blob == NULL) {
+				g_prefix_error(error,
+					       "not valid bytes for %s: ",
+					       fu_genesys_fw_type_to_string(fw_type));
+				return FALSE;
+			}
+			if (!fu_firmware_parse(firmware_sub, blob, flags, error))
+				return FALSE;
```

Comment:
not sure how expensive this is for you, but you can do `fu_firmware_parse_full(firmware_sub, fw, fw_offset, flags, error))` and avoid the fu_bytes_new_offset() call.